### PR TITLE
Building the Visitor Frontend for each Stack Deployed

### PR DESCRIPTION
**Pull request by Mjtlittle**
I stood up my own pipeline and am building two stages of the full makerspace stack. This pull request is just to pass the idea by everyone before I adapt it to the production pipeline. This does not solve the dev-stack or the lambda-CORS aspects of this issue.

The two deployments, `DevA` and `DevB`, should make requests to `https://a` and `https://b` respectivly.

- DevA: https://d87p3t5nf9eu5.cloudfront.net
- DevB: https://dh5ajzrdzv4ys.cloudfront.net

This is just a proof of concept, and is not as clean as I would like. I have spent alot of time looking into potentially sticking the build stage into the stack itself within another pipeline, but this proved overly complicated and other efforts were not as clear.

### CODE REVIEW

**Comment by Mjtlittle** [comment]
Forgot to mention, but this also adds a manual review pre-req for the `DevB` stack stage

File: cdk/DevPipeline.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+from makerspace import MakerspaceStage
+from accounts_config import accounts
+
+class DevPipeline(core.Stack):
+    def __init__(self, app: core.App, id: str, *,
+                 env: core.Environment) -> None:
+        super().__init__(app, id, env=env)
+
+        deploy_cdk_shell_step = ShellStep("Synth",
+            input=CodePipelineSource.connection("Mjtlittle/unified-makerspace", "pipeline-development",
```
Comment (dggorsen) [inline cdk/DevPipeline.py:15]: Just to be clear- this is just for feedback, and wont be pushed, correct?

Comment (Mjtlittle) [reply cdk/DevPipeline.py:15]: Correct! I plan on revising the PR tonight to work with the actual pipeline

_*edit: i'm gonna open a new pr, so I can keep the testing stuff here_

File: cdk/DevPipeline.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                connection_arn="arn:aws:codestar-connections:us-east-1:577283310524:connection/2e5ad00f-c205-4b57-a752-6aa33d55236c"
+            ),
+            commands=[
```
Comment (dggorsen) [inline cdk/DevPipeline.py:18]: I could see value in trying to add multiple build steps and decoupling the logic for each of these 3.

File: cdk/DevPipeline.py (hunk 1 above)
Comment (dggorsen) [reply cdk/DevPipeline.py:15]: Sweet!

File: cdk/DevPipeline.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            synth=deploy_cdk_shell_step,
+            cross_account_keys=True
+        )
+        
+        self.a = MakerspaceStage(self, 'DevA', env=accounts['Dev-jlittl8'])
+        pipeline.add_stage(self.a)
+
+        self.b = MakerspaceStage(self, 'DevB', env=accounts['Dev-jlittl8'])
+        pipeline.add_stage(self.b,
+            pre=[ManualApprovalStep("PromoteToDevB")]
```
Comment (dggorsen) [inline cdk/DevPipeline.py:55]: Love love love to see this!

**Review by dggorsen — APPROVED** [review]
Overall, I dont see any reason that this approach wouldn't work. I do believe there are improvements that *could* be made by leveraging features native to code pipelines in order to make this a more extensible/robust solution e.g. multiple build stages with distinct output artifacts and build action environment variables. Having said that, there is a big distinction between cleanliness and extensibility requirements of runtime code vs build time code, and I think you have captured the spirit of what we were trying to accomplish here :D Awesome work diving deep on this! Code Pipelines is a beast to wrap your head around.

**Comment by Mjtlittle** [comment]
> I do believe there are improvements that _could_ be made by leveraging features native to code pipelines in order to make this a more extensible/robust solution e.g. multiple build stages with distinct output artifacts and build action environment variables.

I am having trouble seeing a nicer picture with our current setup. The synth step for the pipeline.CodePipeline generates all of the artifacts (now including the frontend deployments with this pr) required to deploy the site. I am having a hard time sticking steps before the synth of the app to get those frontend files in place **statically**, and ultimately before cloudformation will pick it up and deploy it.

Ive thought of adding a pre step to build the site before the stage is deployed, but this will occur after the synth of the stack, thus preventing me from adding it at cf deploy time. With this approach I could provide a bucket to the stack for it to use as the frontend, and then maybe have a codebuild stage before the deploy to create said buckets. But this feels like too much orchestration.

Another Idea I had was maybe populating the bucket after the deploy of the stacks. Like adding another stage/step after deploy, to build the site with the correct env vars and then moving it into the bucket that will be served. After a quick look, this project seems to do something like this: (build-frontend and populate-site-s3 after cdk synth and deploy with gh-actions) https://github.com/jviloria96744/react-aws-cdk-template/blob/master/.github/workflows/main.yml. But this still feels just as decoupled.

An idea I had, that im pretty sure is not efficient (basically the codepipeline for each stack approach), could be using another one of aws services like Amplify to host just the frontend. This would allow us to pass in args for the deployment of the site before any future build during cf synth.

But this is all to say i'm in sortuv a rut (service topology-wise) passed this pr, not sure if im missing some key piece of information either.

File: cdk/DevPipeline.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                # deploy cdk
+                "cd cdk",
+                "npm install -g aws-cdk && pip install -r requirements.txt",
+                "cdk synth"
+            ],
+            primary_output_directory="cdk/cdk.out",
+        )
+
+        deploy_cdk_shell_step.add_output_directory("cdk/visit/console/DevA")
+        deploy_cdk_shell_step.add_output_directory("cdk/visit/console/DevB")
```
Comment (Mjtlittle) [inline cdk/DevPipeline.py:43]: Pretty sure these are unnecessary upon more reading; as all of the statically included assets (`aws_s3_deployment.Source.asset(...`) will be included in the `cdk/cdk.out` folder (as `cdk/cdk.out/asset.*/`)

Comment (dggorsen) [reply cdk/DevPipeline.py:43]: Makes sense to me! Cant hurt to remove for now.

**Review by dggorsen — COMMENTED** [review]
Lets sync tomorrow to see when we can merge this. I think its in pretty good shape to merge but I want to make sure we are both around to follow this through the pipeline and manually verify everything is working like we want it to in beta before enabling it in prod since we dont have any automated testing.

Side note- really great work diving deep on this! This was an ambiguous and complex task

**Comment by Mjtlittle** [comment]
The actual changes for this have been moved to #92

**Pull request closed without merging**